### PR TITLE
Implement Main menu background mode setting

### LINKED ISSUE #5516: Beatmap backgrounds in the main menu
We might want to add an ability to allow players select beatmap backgrounds instead of default in the main menu.
Since we have an ability to skin that, we can make a dropdown in settings like

Main menu background
- default
- skin
- beatmap
- seasonal (in the future)

Some elements like "skin" and "beatmap" will be visible for supporters only (or not).

### PR DESCRIPTION
**Pull request by EVAST9919**
Fixes https://github.com/ppy/osu/issues/5516

### CODE REVIEW

File: osu.Game/Overlays/Settings/Sections/Graphics/MainMenuSettings.cs [hunk 1]
```diff
@@ -2,25 +2,55 @@
...
                     Bindable = config.GetBindable<bool>(OsuSetting.MenuParallax)
                 },
+                backgroundDropdown = new SettingsEnumDropdown<MainMenuBackgroundMode>
+                {
+                    LabelText = "Main menu background",
+                    Bindable = config.GetBindable<MainMenuBackgroundMode>(OsuSetting.MenuBackgroundMode)
+                }
             };
+
+            api.LocalUser.BindValueChanged(user => onUserChanged(user.NewValue), true);
```
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/Graphics/MainMenuSettings.cs]:
```suggestion
            api.LocalUser.BindValueChanged(onUserChanged, true);
```

we prefer the `valuechanged` event to be handled at the destination.

File: osu.Game/Overlays/Settings/Sections/Graphics/MainMenuSettings.cs [hunk 2]
```diff
@@ -2,25 +2,55 @@
...
+        }
+
+        private void onUserChanged(User user)
```
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/Graphics/MainMenuSettings.cs]:
```suggestion
        private void onUserChanged(ValueChangedEvent<User> user)
```

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 3]
```diff
@@ -22,17 +23,20 @@ public class BackgroundScreenDefault : BackgroundScreen
...
-        private Bindable<User> user;
         private Bindable<Skin> skin;
+        private readonly Bindable<WorkingBeatmap> beatmap = new Bindable<WorkingBeatmap>();
+        private readonly Bindable<MainMenuBackgroundMode> backgroundMode = new Bindable<MainMenuBackgroundMode>();
 
         [BackgroundDependencyLoader]
-        private void load(IAPIProvider api, SkinManager skinManager)
+        private void load(SkinManager skinManager, OsuConfigManager config, Bindable<WorkingBeatmap> workingBeatmap)
         {
-            user = api.LocalUser.GetBoundCopy();
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: why is this removed? no test coverage to show that this is a regression is a bit of a worry.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs (hunk 3 above)
Comment (EVAST9919) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: soo should I just move it back? But we already have supporter check in the settings

Comment (peppy) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: What happens if you logout or change user?

Comment (EVAST9919) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: It will change the background mode to default and block the dropdown

Comment (peppy) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: What if the settings overlay isn't loaded? There's no guarantee it will be.

Also I'm pretty sure that logic you have will crash if you go from non-supporter to non-supporter (it will try and change the bindable value while already in a disabled state).

Yes, that logic should probably stay there, but it should not reset the setting nor should it be responsible for governing the setting.

Comment (EVAST9919) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: Yeah, looks/feels better. And there's no setting hard reset anymore

**Comment by peppy** [comment]
I wouldn't say no to having some test coverage of the state changes here. An important consideration is to avoid reloading/replacing the background if it hasn't changed (similar to how `BackgroundScreen` works). I can't be sure if this is working correctly with your current code.

**Comment by EVAST9919** [comment]
> I wouldn't say no to having some test coverage of the state changes here. An important consideration is to avoid reloading/replacing the background if it hasn't changed (similar to how `BackgroundScreen` works). I can't be sure if this is working correctly with your current code.

hmm, yeah, I see some cases

**Comment by peppy** [comment]
still hoping to see a visual test appear here

**Comment by peppy** [comment]
Tests need asserts to make sure the background changes (and does not change when not necessary).

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 4]
```diff
@@ -48,9 +59,57 @@ private void display(Background newBackground)
             currentDisplay++;
         }
 
+        private void onUserChanged(ValueChangedEvent<User> user)
+        {
+            if (BackgroundMode.Value == MainMenuBackgroundMode.Default)
```
Comment (UselessToucan) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: These two if statements can be merged into `BackgroundMode.Value == MainMenuBackgroundMode.Default || user.OldValue?.IsSupporter == user.NewValue?.IsSupporter`

Comment (UselessToucan) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: Also, you can reverse the condition and call `UpdateBackground` when it is true

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 5]
```diff
@@ -48,9 +59,57 @@ private void display(Background newBackground)
...
+            UpdateBackground();
+        }
+
+        private Texture skinBackground(Skin skin) => skin.GetTexture("menu-background");
+
+        private bool skinHasBackground => skin.Value.GetTexture("menu-background") != null;
+
+        private void onSkinChanged(ValueChangedEvent<Skin> skin)
+        {
+            if (BackgroundMode.Value != MainMenuBackgroundMode.Skin)
```
Comment (UselessToucan) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: Same as above, these conditions can be merged and reversed.

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 6]
```diff
@@ -48,9 +59,57 @@ private void display(Background newBackground)
...
+
+            if (skinBackground(skin.OldValue) == skinBackground(skin.NewValue))
+                return;
+
+            UpdateBackground();
+        }
+
+        private void onBeatmapChanged(ValueChangedEvent<WorkingBeatmap> beatmap)
+        {
+            if (BackgroundMode.Value == MainMenuBackgroundMode.Skin && skinHasBackground)
```
Comment (UselessToucan) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: I suggest to call `UpdateBackground` on some condition instead of performing `return`

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 7]
```diff
@@ -48,9 +59,57 @@ private void display(Background newBackground)
...
+                return;
+
+            UpdateBackground();
+        }
+
+        private void onModeChanged(ValueChangedEvent<MainMenuBackgroundMode> mode)
+        {
+            if (mode.OldValue == MainMenuBackgroundMode.Default && mode.NewValue == MainMenuBackgroundMode.Skin && !skinHasBackground)
```
Comment (UselessToucan) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: As above, can be merged and reversed.

**Comment by EVAST9919** [comment]
Not sure if one long statement is better than few short ones with `return`

**Comment by EVAST9919** [comment]
Now it's possible to test all the possible cases and there's a visual response whether background has been updated

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 8]
```diff
@@ -60,8 +119,24 @@ private Background createBackground()
         {
             Background newBackground;
 
-            if (user.Value?.IsSupporter ?? false)
-                newBackground = new SkinnedBackground(skin.Value, backgroundName);
+            if (User.Value?.IsSupporter ?? false)
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: this is ugly to read and has code duplication. you're looking for something like this

image

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 9]
```diff
@@ -22,23 +26,36 @@ public class BackgroundScreenDefault : BackgroundScreen
 
         private string backgroundName => $@"Menu/menu-background-{currentDisplay % background_count + 1}";
 
-        private Bindable<User> user;
-        private Bindable<Skin> skin;
+        public readonly Bindable<MainMenuBackgroundMode> BackgroundMode = new Bindable<MainMenuBackgroundMode>();
+        public Bindable<User> User = new Bindable<User>();
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: private

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 10]
```diff
@@ -22,23 +26,36 @@ public class BackgroundScreenDefault : BackgroundScreen
...
+        public Bindable<Skin> Skin = new Bindable<Skin>();
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: private

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 11]
```diff
@@ -22,23 +26,36 @@ public class BackgroundScreenDefault : BackgroundScreen
...
+
+        private readonly Bindable<WorkingBeatmap> beatmap = new Bindable<WorkingBeatmap>();
+        private WorkingBeatmap currentBeatmap;
 
         [BackgroundDependencyLoader]
-        private void load(IAPIProvider api, SkinManager skinManager)
+        private void load(IAPIProvider api, SkinManager skinManager, OsuConfigManager config, Bindable<WorkingBeatmap> workingBeatmap)
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]:
```suggestion
        private void load(IAPIProvider api, SkinManager skinManager, OsuConfigManager config, Bindable<WorkingBeatmap> beatmap)
```

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 12]
```diff
@@ -22,23 +26,36 @@ public class BackgroundScreenDefault : BackgroundScreen
...
         {
-            user = api.LocalUser.GetBoundCopy();
-            skin = skinManager.CurrentSkin.GetBoundCopy();
-
-            user.ValueChanged += _ => Next();
-            skin.ValueChanged += _ => Next();
+            Skin = skinManager.CurrentSkin.GetBoundCopy();
+            beatmap.BindTo(workingBeatmap);
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: why using `GetBoundCopy` for one but not the other?

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 13]
```diff
@@ -48,9 +65,51 @@ private void display(Background newBackground)
             currentDisplay++;
         }
 
-        private ScheduledDelegate nextTask;
+        private Texture skinBackground(Skin skin) => skin.GetTexture("menu-background");
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: what is this even

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 14]
```diff
@@ -48,9 +65,51 @@ private void display(Background newBackground)
...
+            if (!((User.Value?.IsSupporter ?? true) && BackgroundMode.Value == MainMenuBackgroundMode.Skin && SkinHasBackground))
+            {
+                UpdateBackground();
+                currentBeatmap = beatmap.NewValue;
+            }
+        }
+
+        private void onModeChanged(ValueChangedEvent<MainMenuBackgroundMode> mode)
+        {
+            if (((User.Value?.IsSupporter) ?? false)
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: too many conditions

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 15]
```diff
@@ -48,9 +65,51 @@ private void display(Background newBackground)
...
+
+        private void onSkinChanged(ValueChangedEvent<Skin> skin)
+        {
+            if ((User.Value?.IsSupporter ?? false) && BackgroundMode.Value == MainMenuBackgroundMode.Skin && skinBackground(skin.OldValue) != skinBackground(skin.NewValue))
+                UpdateBackground();
+        }
+
+        private void onBeatmapChanged(ValueChangedEvent<WorkingBeatmap> beatmap)
+        {
+            if (!((User.Value?.IsSupporter ?? true) && BackgroundMode.Value == MainMenuBackgroundMode.Skin && SkinHasBackground))
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: too many conditions

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 16]
```diff
@@ -48,9 +65,51 @@ private void display(Background newBackground)
...
+        private void onUserChanged(ValueChangedEvent<User> user)
+        {
+            if (BackgroundMode.Value != MainMenuBackgroundMode.Default && user.OldValue?.IsSupporter != user.NewValue?.IsSupporter &&
+                ((BackgroundMode.Value != MainMenuBackgroundMode.Skin) | SkinHasBackground))
+                UpdateBackground();
+        }
+
+        private void onSkinChanged(ValueChangedEvent<Skin> skin)
+        {
+            if ((User.Value?.IsSupporter ?? false) && BackgroundMode.Value == MainMenuBackgroundMode.Skin && skinBackground(skin.OldValue) != skinBackground(skin.NewValue))
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: too many conditions

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 17]
```diff
@@ -48,9 +65,51 @@ private void display(Background newBackground)
...
         }
 
-        private ScheduledDelegate nextTask;
+        private Texture skinBackground(Skin skin) => skin.GetTexture("menu-background");
+
+        protected bool SkinHasBackground => Skin.Value.GetTexture("menu-background") != null;
+
+        private void onUserChanged(ValueChangedEvent<User> user)
+        {
+            if (BackgroundMode.Value != MainMenuBackgroundMode.Default && user.OldValue?.IsSupporter != user.NewValue?.IsSupporter &&
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: too many conditions

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 18]
```diff
@@ -48,9 +65,51 @@ private void display(Background newBackground)
...
+        {
+            if (((User.Value?.IsSupporter) ?? false)
+                && (!(mode.OldValue == MainMenuBackgroundMode.Default && mode.NewValue == MainMenuBackgroundMode.Skin && !SkinHasBackground)
+                    & !(mode.OldValue == MainMenuBackgroundMode.Skin && mode.NewValue == MainMenuBackgroundMode.Default && !SkinHasBackground)))
+                UpdateBackground();
+        }
 
         public void Next()
+        {
+            if (BackgroundMode.Value == MainMenuBackgroundMode.Default ||
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: too many conditions

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 19]
```diff
@@ -86,5 +161,22 @@ private void load()
                 Sprite.Texture = skin.GetTexture("menu-background") ?? Sprite.Texture;
             }
         }
+
+        private class BeatmapBackground : Background
+        {
+            private readonly WorkingBeatmap beatmap;
+
+            public BeatmapBackground(WorkingBeatmap beatmap, string fallbackTextureName)
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: fallback is not going to work as you expect. check the base code - it's *always* going to load it.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs [hunk 20]
```diff
@@ -22,23 +26,36 @@ public class BackgroundScreenDefault : BackgroundScreen
...
             currentDisplay = RNG.Next(0, background_count);
 
             display(createBackground());
         }
 
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            User.BindValueChanged(onUserChanged, true);
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: i'm 99.9% sure this is going to trigger multiple background reloads on first run.

File: osu.Game/Screens/Menu/MainMenu.cs [hunk 21]
```diff
@@ -162,14 +158,6 @@ protected override void LogoSuspending(OsuLogo logo)
             seq.OnAbort(_ => buttons.SetOsuLogo(null));
         }
 
-        private void beatmap_ValueChanged(ValueChangedEvent<WorkingBeatmap> e)
-        {
-            if (!this.IsCurrentScreen())
```
Comment (peppy) [inline osu.Game/Screens/Menu/MainMenu.cs]: you removed this conditional. it's going to be triggering changes even when the background isn't current?

File: osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs (hunk 9 above)
Comment (EVAST9919) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: I made them public for test purposes

Comment (peppy) [reply osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs]: At most we use protected for testing, but looking at your tests I believe they are not required to be public (you should be changing the DI sources instead)

**Comment by EVAST9919** [comment]
Needs total rewrite because the code quality leaves much to be desired.
